drivers/pn532/pn532.c

```diff
@@ -0,0 +1,639 @@
+        pos = 5;
+
+    }
+    else {
+        buff[3] = 0xff;
+        buff[4] = 0xff;
+        buff[5] = (len >> 8) & 0xff;
+        buff[6] = (len) & 0xff;
+        buff[7] = 0x00 - buff[5] - buff[6];
+        pos = 8;
+    }
+
+    buff[pos] = 0xD4;
+    checksum = chksum(&buff[pos], len);
+
+    len += pos;
+    buff[len++] = checksum;
+    buff[len++] = 0x00;
+
+    return _write(dev, buff, len);
+}
+
+static void wait_ready(pn532_t *dev)
+{
+    mutex_lock(&dev->trap);
+}
+
+/* Returns >0 payload len (or <0 received len but not as expected */
+static int read_command(pn532_t *dev, char *buff, unsigned len, int expected_cmd)
+{
+    int r;
+    unsigned j, fi, lp, lc;
+
+    /* apply framing overhead */
+    len += 8;
+    if (len >= 0xff) {
+        len += 3;
+    }
+
+    r = _read(dev, buff, len);
+
+    /* Validate frame structure and CRCs
+     *
+     * Note that all offsets are shifted by one since the first byte is always
+     * 0x01. */
+    if (r < len || buff[1] != 0x00 || buff[2] != 0x00 || buff[3] != 0xFF
+        || buff[r - 1] != 0x00) {
+        return -r;
+    }
+
+    if (buff[4] == 0xff && buff[5] == 0xff) {
+        /* extended frame */
+        lp = buff[6] << 8 | buff[7];
+        lc = (buff[6] + buff[7] + buff[8]) & 0xff;
+        fi = 9;
+
+    }
+    else {
+        /* normal frame */
+        lp = buff[4];
+        lc = (buff[4] + buff[5]) & 0xff;
+        fi = 6;
+    }
+
+    if (lc != 0 || lp >= 265 || buff[fi] != 0xD5) {
+        return -r;
+    }
+
+    if (chksum(&buff[fi], lp) != buff[fi + lp]) {
+        return -r;
+    }
+
+    if (buff[fi + 1] != expected_cmd) {
+        return -r;
+    }
+
+    /* Move the meaningful data to the beginning of the buffer */
+    /* start copying after command byte */
+    for (j = 0, fi += 2, lp -= 2; j < lp; fi++, j++) {
+        buff[j] = buff[fi];
+    }
+
+    DEBUG("pn532: in cmd ");
+    PRINTBUFF(buff, lp);
+
+    return lp;
+}
+
+/* Returns 0 if OK, <0 otherwise */
+static int send_check_ack(pn532_t *dev, char *buff, unsigned len)
+{
+    static char ack[] = { 0x00, 0x00, 0xff, 0x00, 0xff, 0x00 };
+
+    if (send_cmd(dev, buff, len) > 0) {
+        wait_ready(dev);
+        if (_read(dev, buff, sizeof(ack)) != sizeof(ack) + 1) {
+            return -2;
+        }
+
+        if (memcmp(&buff[1], ack, sizeof(ack)) != 0) {
+            return -3;
+        }
+
+        wait_ready(dev);
+        return 0;
+    }
+
+    return -1;
+}
+
+/* sendl: send length, recvl: receive payload length */
+static int send_rcv(pn532_t *dev, char *buff, unsigned sendl, unsigned recvl)
+{
+    int expected_cmd = buff[BUFF_CMD_START] + 1;
+
+    if (send_check_ack(dev, buff, sendl + 1)) {
+        return 0;
+    }
+
+    recvl += 1; /* cmd response */
+    return read_command(dev, buff, recvl, expected_cmd);
+}
+
+int pn532_fw_version(pn532_t *dev, uint32_t *fw_ver)
+{
+    unsigned ret = -1;
+    char buff[PN532_BUFFER_LEN];
+
+    buff[BUFF_CMD_START     ] = CMD_FIRMWARE_VERSION;
```

Comment:
They are there for consistency with the other assignments to the same buffer variable. Maybe if there's a single byte it doesn't make much sense...